### ISSUE TITLE
Feature request: Sync from local node with BIP157 filters

### ISSUE THREAD

**Issue by andrewtoth** [spesmilo/electrum#5912]
Currently electrum requires an address index to sync, which requires users to trust third party servers that have this index or use ElectrumX, electrs, or EPS. With compact block filters Electrum can be synced without this index in reasonable time. This would let users connect to their own node without trusting any intermediate software.

**Comment by ecdsa** [comment]
We are aware of BIP157. However, opening an issue about it does not really add any useful value to us, unless you have insight to share about how precisely it would interface with the Electrum codebase

**Comment by andrewtoth** [comment]
Hi @ecdsa 

I couldn't find any other issues about it, so I thought it would be a good idea to open one to track any discussion. Would you have any useful links to discussion/docs about supporting this feature? Do you know if anyone is currently working on this and what the major challenges would be to add this feature? I'd be interested to help add this feature but wanted to first see what the current situation is.

Thanks

**Comment by ecdsa** [comment]
nobody is working on adding that feature at the moment.

**Comment by chris-belcher** [comment]
As of April 2019 the size of all the bip158 filters is 4GB [1]. If the user is in a situation where disk space is scarce they'd be better off using EPS and enabling pruning. If the user has lots of disk space they'd use ElectrumX or electrs.

BIP158 filters would also presumably have to be downloaded by the wallet. So imagine a user's full node running on a raspberry pi and the user's smartphone has to download 4GB from it. This is a much higher bandwidth requirement than the status quo, such a wallet can't really be called "lightweight".

The point about needing to trust intermediate software isn't valid, because all that software is open source.

Because of these reasons I don't really see the value added by such filters. 

[1] https://bitcoinops.org/en/newsletters/2019/04/23/

**Comment by andrewtoth** [comment]
@chris-belcher I didn't say they'd have to trust intermediate software, I said they could either trust third party servers *or* have to run intermediate software. I think the option of not having to use something to glue Bitcoin Core and Electrum together would be very valuable. Giving Electrum the rpcuser and password and then having it just work would be great. Just having to use intermediate software adds a lot of friction to the experience, not to mention dealing with issues and maintenance.

That's a good point regarding bandwidth cost. Currently the block filters take up 4.7 GB on my node. We can skip some of this with a birthdate or start at block 481824 if it's a segwit wallet. Also, these only have to be downloaded once. For a new wallet and after syncing an old wallet it becomes much lighter.

**Comment by Talkless** [comment]
> Just having to use intermediate software adds a lot of friction to the experience

Yes that's my view too.

For most advanced users, they can run ElectrumX server (together with bitcoind of course) and provide fast user expereince for other simple Electrum users via P2P protocol.

For most simple users, they just launch Electrum and use these P2P servers, if they don't care (or don't know) about privacy issues.

Now, this solution (get block filters, fee info, more?) via RPC would be a middle ground. You can use you own full node, but no more intermediate software are needed. Since you waited untill your full node was synced, waiting for Electrum to scan all block filters is not much a cost, really. It's still better than syncing ElectrumX server (for a month in my Odroid-HC1 case)...

**Comment by SomberNight** [comment]
_(just something I can link to)_

BIP157/158 vs using Electrum server:
- (pro) better privacy: server currently learns all wallet addresses; with BIP157/158 it would not learn any
- (contra) much slower initial sync time when restoring wallet
    - either slow for every wallet restore, or slow only for the first one but then the client needs to store the filters (several GB)
- (contra) somewhat higher ongoing bandwidth usage
- (contra) no mempool: clients would not get notifications for unconfirmed transactions
    - incoming unconfirmed txn cannot be CPFP-ed by client as client does not see it

**Comment by andrewtoth** [comment]
@SomberNight I think another pro is not having to deal with setting up Electrum server at all.

**Comment by SomberNight** [comment]
@andrewtoth I've clarified the comparison now. It depends on whether the user runs their own bitcoind/server, in which case the privacy benefit goes away (already perfect privacy).

**Comment by inaltoasinistra** [comment]
@SomberNight if Electrum is connected to a user's own bitcoind the mempool is accessible via rpc or zmq

**Comment by shesek** [comment]
> Giving Electrum the rpcuser and password and then having it just work would be great. Just having to use intermediate software adds a lot of friction to the experience ...  not having to deal with setting up Electrum server at all 

Apologies for the shameless self-plug, but @andrewtoth if the friction of intermediate software is what bothers you, you might be interested in checking out the [bwt plugin](https://github.com/bwt-dev/bwt-electrum-plugin).

**Comment by Talkless** [comment]
@shesek Yeah I know bwt, but since it's not available by default in Electrum, I would still categorize it as "intermediate software", i.e. something you will have to download, install, and maintain (update) manually, _in addition_ to the Electrum.
